File: server/carrier_move.go
package main

// active_player_ai, check_collectibles, find_route_left/right_first.
// Called after hardware targeting; nil means enter the role's pass/shot branch.
func (s *State) carrierMove(i, random int, distances *[18]int) *passPlan {
	if i%9 == 0 {
		return nil
	}
	const unit = 22.4 / 576
	p := &s.Players[i]
	j := s.Controlled[1-p.Team]
	if j >= 0 && s.Players[j].Stun <= 0 && s.Players[j].Health > 0 && distances[j] <= 64 {
		return nil
	}
	blocked := s.opponentDirections(i, distances)
	clear := func(x, z float64) bool { key := passDirection(p, x, z); return key != blocked[0] && key != blocked[1] }
	intelligence := p.Stats[7]
	if i%9 >= 6 {
		intelligence /= 2
	}
	if intelligence > random {
		for _, slot := range []int{0, 1, 6, 2, 3, 4, 5} {
			item := s.Pickups[slot]
			if item.Kind != 0 && item.Wait <= 0 && s.worldInViewport(item.X, item.Z, 0) && clear(item.X, item.Z) {
				return &passPlan{x: item.X, z: item.Z}
			}
		}
	}
	side := p.Team
	if s.Period == 2 {
		side ^= 1
	}
	zone := supportZones[side][i%9]
	xmin, xmax, ymin, ymax := zone[0], zone[1], zone[2], zone[3]
	cy := ((ymin + ymax) / 2) &^ 1
	x, y := supportTerrain(*p)
	if side == 0 && y <= cy || side == 1 && y >= cy {
		return nil
	}
	forwardX := max(xmin, min(xmax, x))
	if forwardX <= 96 {
		forwardX = 56
	} else if forwardX >= 544 {
		forwardX = 584
	}
	forwardX = max(xmin, min(xmax, forwardX))
	frontEdge, backEdge := ymin, ymax
	if side != 0 {
		frontEdge, backEdge = ymax, ymin
	}
	front := (y + frontEdge) / 2
	lateral := [2]int{xmax, xmin}
	if random&8 != 0 {
		lateral = [2]int{xmin, xmax}
	}
	candidates := [][2]int{{forwardX, front}}
	for _, targetY := range []int{front, (y + cy) / 2, (y + backEdge) / 2} {
		for _, edge := range lateral {
			candidates = append(candidates, [2]int{(x + edge) / 2, targetY})
		}
	}
	for _, point := range candidates {
		x, z := float64(576-point[1])*unit, float64(point[0]-320)*unit
		if clear(x, z) {
			return &passPlan{x: x, z: z}
		}
	}
	return nil
}

File: src/carrier-move.js
import { zones } from "./support-data.js";
import { directionKey, opponentDirections } from "./defensive-pass.js";
import { worldInViewport } from "./visibility.js";
const unit = 22.4 / 576;

// active_player_ai, check_collectibles, find_route_left/right_first.
// Called after hardware targeting; null means enter the role's pass/shot branch.
export function carrierMove(s, i, random, distances) {
  if (i % 9 === 0) return null;
  const p = s.players[i],
    j = s.controlled[1 - p.team],
    q = s.players[j];
  if (q && q.stun <= 0 && q.health > 0 && distances[j] <= 64) return null;
  const blocked = opponentDirections(s, i, distances);
  const clear = (target) =>
    !blocked.includes(directionKey(p, target.x, target.z));
  const intelligence = i % 9 >= 6 ? Math.floor(p.stats[7] / 2) : p.stats[7];
  if (intelligence > random) {
    for (const slot of [0, 1, 6, 2, 3, 4, 5]) {
      const item = s.pickups?.[slot];
      if (item && item.kind && item.wait <= 0 && worldInViewport(s, item) && clear(item))
        return { x: item.x, z: item.z };
    }
  }
  const side = p.team ^ (s.period === 2 ? 1 : 0);
  const [xmin, xmax, ymin, ymax] = zones[side][i % 9];
  const cy = Math.floor((ymin + ymax) / 2) & ~1;
  const x = Math.round(320 + p.z / unit),
    y = Math.round(576 - p.x / unit);
  if (side === 0 ? y <= cy : y >= cy) return null;
  const bound = (v) => Math.max(xmin, Math.min(xmax, v));
  let forwardX = bound(x);
  if (forwardX <= 96) forwardX = 56;
  else if (forwardX >= 544) forwardX = 584;
  forwardX = bound(forwardX);
  const front = Math.floor((y + (side === 0 ? ymin : ymax)) / 2);
  const lateral = random & 8 ? [xmin, xmax] : [xmax, xmin];
  const candidates = [[forwardX, front]];
  for (const targetY of [
    front,
    Math.floor((y + cy) / 2),
    Math.floor((y + (side === 0 ? ymax : ymin)) / 2),
  ]) {
    for (const edge of lateral)
      candidates.push([Math.floor((x + edge) / 2), targetY]);
  }
  for (const [tx, ty] of candidates) {
    const target = { x: (576 - ty) * unit, z: (tx - 320) * unit };
    if (clear(target)) return target;
  }
  return null;
}
